### ISSUE TITLE
Support Mypy Type Checking

### ISSUE THREAD

**Issue by johnthagen** [drhagen/parsita#37]
When adding `parsita` to a project that is type checked with `mypy`, we had to add the following configuration to our `mypy` configuration:

```toml
[tool.mypy]
strict = true
disallow_subclassing_any = false
```

Note that `disallow_subclassing_any` is not enabled by default, but we use `strict = true` above to provide better type checking in our Python projects.

Otherwise we get a lot of errors of the form:

```
 error: Class cannot subclass "TextParsers" (has type "Any")  [misc]
```

I'm not sure if this is something that can be fixed upstream in `parsita`, but at a minimum having a type checking page in the documentation with known issues like this could help future users.

### Environment

- Python 3.9
- `parsita` 1.7.1
- mypy 0.982

**Comment by drhagen** [comment]
Is this a problem because Parsita is not marked as `py.typed` or is this due to something uniquely weird about Parsita? I just want to know if this can be fixed with something other than full mypy compatibility.

**Comment by johnthagen** [comment]
Good point, I didn't check that parista is missing `py.typed`. That may solve the issue or possibly reveal other typing issues. That would be the first step to figure out if mypy could be used.

**Comment by drhagen** [comment]
I took a look at bringing Parsita up to mypy's standards. It is doable, but it's a non-trivial amount of work, and I don't have any experience doing this. To anyone reading this: I would accept a PR to make Parsita mypy compatible.

**Comment by AndrewMHenry** [comment]
I would add a vote for mypy support. It's easy enough to work around by explicitly suppressing `parsita` errors, but it would be nice not to need that.

Also, @drhagen, thanks so much for all your work on this library; it's really slick and fun to use!

**Comment by drhagen** [comment]
I guess I haven't taken a look at this since the release of Parsita 2.0. A lot of things got simplified in the major release, so maybe it will be doable this time around.

**Comment by drhagen** [comment]
After spending a ton of time playing around with Python types in Parsita, I have concluded that it is basically impossible to do well. I have made a best-effort at typing and add `py.typed`. There are lots of `Any`s and lots of signatures that are insufficiently specific, but mypy thinks they are all correct now.

Python lacks [lower type bounds](https://github.com/python/typing/issues/674) and [higher-kinded types](https://github.com/python/typing/issues/548). As such, it is unable to express many of the operations that a parser combinator library must perform.
